Fix overlapping infantry and building decorations

pchote opened the pull request:
This PR polishes our overlay decorations for infantry and structures.  The repair and powerdown icons are now shown side by side in RA/D2K/TS, and the heal/tiberium protection/veterancy icons now alternate on infantry where there is not enough space to do anything else.

This also adds the hospital heal indicator to RA, and changes it to blink and only display when the infantry is actively healing.  The hazmat indicator in TD is similarly only shown when the infantry is in a tiberium field.

Supersedes #14810.
Closes #14736.

pchote committed:
Document RepairableBuilding fields.

pchote committed:
Add conditional offset support to WithDecoration.

pchote committed:
Fix overlapping repair / powerdown indicators.

pchote committed:
Replace repair indicator effect with a Decoration subclass.

pchote commented:
Fixed the code style warnings.

GraionDilach commented:
The yaml syntax looks really confusing on first sight.

reaperrr commented on mods/ra/rules/defaults.yaml:
```diff
@@ -145,6 +145,24 @@
+	WithDecoration@RANK-1:
+		BlinkInterval: 32
+		BlinkPatterns:
+			hospitalheal: True, False
```
This is really bound to confuse modders. In my opinion it would be more modder-friendly to introduce a `Blink` enum with `On`/ `Off` and map that to True/False internally.

pchote replied:
What if I rename the fields to `VisibilityInterval`, `VisibilityPattern`, `VisibilityPatterns` instead?

GraionDilach replied:
Wouldn't help, Blink sounds more reasonable here. I grasped the syntax now and tbh I don't think renaming this would help. It needs more detailed descriptions IMO.

pchote committed:
Add blink parameters to WithDecoration.

pchote committed:
Fix overlapping veterancy/hospital/hazmat indicators in TD.

This also changes the hazmat indicator to only be visible when on
tiberium, and the hospital indicator to blink while active and only
be visible when the infantry is damaged.

pchote committed:
Add hospital heal indicators to RA.

pchote committed:
Fix experience/hospital indicator overlap in TS.

pchote commented:
Added blink On/Off enum.

GraionDilach commented:
I would open an issue after this PR gets merged to refactor RepairableBuildings to utilize the stacked conditions for the repair effect as well.

GraionDilach reviewed (dismissed):
:+1: otherwise.

pchote commented:
Rebased.

Smittytron reviewed (dismissed):
Looked good in testing

reaperrr commented:
Needs another rebase, LGTM after that.

pchote commented:
Rebased.

pchote commented on OpenRA.Mods.Common/UpdateRules/UpdatePath.cs:
```diff
@@ -53,7 +53,8 @@ public class UpdatePath
 				new DefineOwnerLostAction(),
-				new RenameEmitInfantryOnSell()
+				new RenameEmitInfantryOnSell(),
+				new SplitRepairDecoration(),
```
I've deliberately added a comma here in the hopes that it will reduce the rebase conflicts going forward.

reaperrr replied:
Makes sense, I'll try to keep that in mind for my PRs as well.

reaperrr commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/506b83167b8da17dc0f94eeba345cdb718cd1876...992f30c8649588aadcbd84b4881dbbf9db4d53bc

dragunoff commented:
I think that the blinking hospital indicator is a bit jarring. IMO it should be simply displayed, without blinking when infantry is healing. If it is to blink, then I'd suggest making it smoothly pulse by animating the opacity of the overlay. 

I'm not sure whether I should file a new issue for that, just my opinion here.
